### PR TITLE
Fix condition in realloc() implementation of Vec

### PR DESCRIPTION
**Pull request by link2xt**
Poolable and Realloc trait method implementations for Vec
operate on the vector itself rather than its underlying buffer.

This is why Poolable::capacity() implementation calls Vec::len()
and Realloc::realloc() implementation uses Vec::resize() rather
than Vec::reserve() or Vec::reserve_exact().

However, the condition checking whether the vector should be resized
called Vec::len() rather than Vec::capacity().
Apparently it was an attempt to call Poolable::capacity(&self),
but by default ambiguous call is resolved to Vec::capacity().

The problem is resolved by calling .len() explicitly.

Regression test and assertions failing for byte-pool 0.2.2 are added.

### CODE REVIEW

**Comment by link2xt** [comment]
cc @flub @dignifiedquire

**Comment by link2xt** [comment]
This `capacity` calling `len` and implemented in a trait on top of `Vec` is very unfortunate:
https://github.com/dignifiedquire/byte-pool/blob/2a50848d9f7924511eda7705e493ea4c8316e23f/src/poolable.rs#L10-L13

This was noticed before (well, yesterday), but never addressed: #4

**Comment by link2xt** [comment]
Once this is merged, byte-pool 0.2.3 is needed to merge https://github.com/async-email/async-imap/pull/77 and then release a new deltachat-core-rust.

**Review by flub — APPROVED** [review]

**Comment by link2xt** [comment]
@flub I pushed a second commit here as you said in https://github.com/async-email/async-imap/pull/77#issuecomment-1498714513 that "byte-pool wanted to never have a Vec with capacity != len since that's wasteful, but at the time of writing this was not yet possible with the rust std api."
Now length is checked to be the same as capacity, which is ensured by using [`reserve_exact`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.reserve_exact) before resize.

File: src/poolable.rs [hunk 1]
```diff
@@ -41,11 +41,21 @@ impl<T: Default + Clone> Realloc for Vec<T> {
...
 
         assert!(new_size > 0);
-        match new_size.cmp(&self.capacity()) {
-            Greater => self.resize(new_size, T::default()),
+        match new_size.cmp(&self.len()) {
+            Greater => {
+                self.reserve_exact(new_size - self.len());
+                debug_assert_eq!(self.capacity(), new_size);
+                self.resize(new_size, T::default());
+                debug_assert_eq!(self.len(), new_size);
```
Comment (flub) [inline src/poolable.rs:49]: this assert seems a bit redundant, but doesn't do much harm either.

**Review by flub — APPROVED** [review]

**Comment by flub** [comment]
Hum, don't you have to do the reserve_exact thing also when initially creating the block?

**Comment by link2xt** [comment]
> Hum, don't you have to do the reserve_exact thing also when initially creating the block?

I have quickly checked, `vec![T::default(); size]` allocates exact size buffer. But even if not and it changes later, it is not a big deal, we look at len() everywhere and matching capacity to vector size is only a memory usage optimization.

**Review by flub — APPROVED** [review]

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**